dev/TeachingTip/TestUI/TeachingTipPage.xaml.cs

```diff
@@ -0,0 +1,589 @@
+        {
+            TeachingTip.SetAttach(this.targetButton, this.TeachingTip);
+        }
+
+        public void OnUntargetButtonClicked(object sender, RoutedEventArgs args)
+        {
+            TeachingTip.SetAttach(null, this.TeachingTip);
+        }
+
+        public void OnShowButtonClicked(object sender, RoutedEventArgs args)
+        {
+            this.TeachingTip.IsOpen = true;
+        }
+
+        public void OnCloseButtonClicked(object sender, RoutedEventArgs args)
+        {
+            this.TeachingTip.IsOpen = false;
+        }
+
+        public void OnSetTargetVerticalAlignmentButtonClicked(object sender, RoutedEventArgs args)
+        {
+            if(TargetVerticalAlignmentComboBox.SelectedIndex == 0)
+            {
+                this.targetButton.VerticalAlignment = VerticalAlignment.Top;
+            }
+            else if (TargetVerticalAlignmentComboBox.SelectedIndex == 1)
+            {
+                this.targetButton.VerticalAlignment = VerticalAlignment.Center;
+            }
+            else
+            {
+                this.targetButton.VerticalAlignment = VerticalAlignment.Bottom;
+            }
+            OnGetTargetBoundsButtonClicked(null, null);
+        }
+        public void OnSetTargetHorizontalAlignmentButtonClicked(object sender, RoutedEventArgs args)
+        {
+            if (TargetHorizontalAlignmentComboBox.SelectedIndex == 0)
+            {
+                this.targetButton.HorizontalAlignment = HorizontalAlignment.Left;
+            }
+            else if (TargetHorizontalAlignmentComboBox.SelectedIndex == 1)
+            {
+                this.targetButton.HorizontalAlignment = HorizontalAlignment.Center;
+            }
+            else
+            {
+                this.targetButton.HorizontalAlignment = HorizontalAlignment.Right;
+            }
+            OnGetTargetBoundsButtonClicked(null, null);
+        }
+
+        public void OnSetAnimationParametersButtonClicked(object sender, RoutedEventArgs args)
+        {
+            var expandEasing = Window.Current.Compositor.CreateCubicBezierEasingFunction(
+                new System.Numerics.Vector2(float.Parse(this.ExpandControlPoint1X.Text), float.Parse(this.ExpandControlPoint1Y.Text)),
+                new System.Numerics.Vector2(float.Parse(this.ExpandControlPoint2X.Text), float.Parse(this.ExpandControlPoint2Y.Text)));
+
+            var contractEasing = Window.Current.Compositor.CreateCubicBezierEasingFunction(
+                new System.Numerics.Vector2(float.Parse(this.ExpandControlPoint1X.Text), float.Parse(this.ExpandControlPoint1Y.Text)),
+                new System.Numerics.Vector2(float.Parse(this.ExpandControlPoint2X.Text), float.Parse(this.ExpandControlPoint2Y.Text)));
+
+            TeachingTipTestHooks.SetExpandEasingFunction(this.TeachingTip, expandEasing);
+            TeachingTipTestHooks.SetContractEasingFunction(this.TeachingTip, contractEasing);
```

Comment:
What is the purpose of these two test hooks?

Reply:
It lets us play with the knobs on the animation without rebuilding the test app.

---
In reply to: [252904845](https://github.com/Microsoft/microsoft-ui-xaml/pull/32#discussion_r252904845) [](ancestors = 252904845)